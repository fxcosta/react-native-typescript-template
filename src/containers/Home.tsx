import Strings from '../constants/strings';
import React, {Component} from 'react';
import {Platform, StyleSheet, Text, View} from 'react-native';
import {connect} from 'react-redux';
import {bindActionCreators, Dispatch} from 'redux';

import * as UserActions from '../store/ducks/users/actions';
import {User} from '../store/ducks/users/types';
import {ApplicationState} from 'src/store';

const instructions = Platform.select({
  ios: 'Press Cmd+R to reload,\n' + 'Cmd+D or shake for dev menu',
  android: 'Double tap R on your keyboard to reload,\n' + 'Shake or press menu button for dev menu',
});

interface StateProps {
  user: User | null;
}

interface DispatchProps {
  fetchUser(payload: string): void;
}

type Props = StateProps & DispatchProps;

export class Home extends Component<Props> {
  componentDidMount() {
    this.props.fetchUser('1');
  }

  render() {
    return (
      <View style={styles.container}>
        <Text style={styles.welcome}>{Strings.hello}. Welcome to React Native!</Text>
        <Text style={styles.instructions}>To get started, edit Home.tsx</Text>
        <Text style={styles.instructions}>{instructions}</Text>
      </View>
    );
  }
}

const mapStateToProps = (state: ApplicationState) => ({
  user: state.users.data,
});

const mapDispatchToProps = (dispatch: Dispatch) => bindActionCreators(UserActions, dispatch);

export default connect(mapStateToProps, mapDispatchToProps)(Home);

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: '#F5FCFF',
  },
  welcome: {
    fontSize: 20,
    textAlign: 'center',
    margin: 10,
  },
  instructions: {
    textAlign: 'center',
    color: '#333333',
    marginBottom: 5,
  },
});
