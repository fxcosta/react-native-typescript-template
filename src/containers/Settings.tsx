import React from 'react';
import {StyleSheet, Text, View} from 'react-native';

export default class Settings extends React.PureComponent {
  render() {
    return (
      <View style={styles.container}>
        <Text> Settings </Text>
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: '#F5FCFF',
  },
});
