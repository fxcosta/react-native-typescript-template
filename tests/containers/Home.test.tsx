import {Home} from '../../containers/Home';
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import React from 'react';
import {render} from 'react-native-testing-library';

Enzyme.configure({adapter: new Adapter()});

const createTestProps = (props?: object) => ({
  fetchUser: jest.fn(),
  user: {},
  ...props,
});

jest.mock('react-redux', () => {
  return {
    connect: jest.fn().mockReturnValue(() => jest.fn()),
  };
});
jest.mock('../../actions/usersActions', () => {
  return {
    fetchUser: jest.fn().mockReturnValue('mock user action'),
  };
});

jest.mock(
  'react-native-localization',
  () =>
    class RNLocalization {
      language = 'en';
      props: any;

      constructor(props: any) {
        this.props = props;
        this.setLanguage(this.language);
      }

      setLanguage(interfaceLanguage: any) {
        this.language = interfaceLanguage;
        if (this.props[interfaceLanguage]) {
          const localizedStrings = this.props[this.language];
          for (const key in localizedStrings) {
            if (localizedStrings.hasOwnProperty(key)) {
              (this as any)[key] = localizedStrings[key];
            }
          }
        }
      }
    },
);

describe('Home', () => {
  const fetchUser = jest.fn();
  const navigate = jest.fn();
  const props: any = createTestProps({
    fetchUser,
    navigation: {
      navigate,
    },
  });
  const {getByText, toJSON} = render(<Home {...props} />);
  it('should render a welcome', () => {
    expect(getByText(/welcome/i)).toBeDefined();
  });

  it('should call fetchUser', () => {
    expect(fetchUser).toBeCalled();
  });

  it('should match snapshot', () => {
    expect(toJSON()).toMatchSnapshot();
  });
});

describe('ConnectedHome', () => {
  const mockConnect = require('react-redux').connect;
  const mapStateToProps = mockConnect.mock.calls[0][0];
  it('should map user from state to props', () => {
    const user = {id: '1'};
    const mockState = {users: {user}};
    const props = mapStateToProps(mockState);

    expect(props.user).toEqual(user);
  });
});
