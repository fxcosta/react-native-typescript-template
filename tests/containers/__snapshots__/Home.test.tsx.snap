// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home should match snapshot 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#F5FCFF",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <Text
    style={
      Object {
        "fontSize": 20,
        "margin": 10,
        "textAlign": "center",
      }
    }
  >
    Hello
    . Welcome to React Native!
  </Text>
  <Text
    style={
      Object {
        "color": "#333333",
        "marginBottom": 5,
        "textAlign": "center",
      }
    }
  >
    To get started, edit Home.tsx
  </Text>
  <Text
    style={
      Object {
        "color": "#333333",
        "marginBottom": 5,
        "textAlign": "center",
      }
    }
  >
    Press Cmd+R to reload,
Cmd+D or shake for dev menu
  </Text>
</View>
`;
